sonar-batch/src/main/java/org/sonar/batch/cache/WSLoader.java

```diff
@@ -55,15 +55,76 @@
   private final LoadStrategy defautLoadStrategy;
   private final ServerClient client;
   private final PersistentCache cache;
+  private final UserProperties userProperties;
   private ServerStatus serverStatus;
 
-  public WSLoader(LoadStrategy strategy, PersistentCache cache, ServerClient client) {
+  private DataLoader<String> stringServerLoader = new DataLoader<String>() {
+    @Override
+    public String load(String id) throws IOException {
+      InputStream is = client.load(id, REQUEST_METHOD, true, CONNECT_TIMEOUT, getReadTimeout());
+      String str = IOUtils.toString(is, StandardCharsets.UTF_8);
+      try {
+        cache.put(id, str.getBytes(StandardCharsets.UTF_8));
+      } catch (IOException e) {
+        throw new IllegalStateException("Error saving to WS cache", e);
+      }
+      return str;
+    }
+
+  };
+
+  private DataLoader<String> stringCacheLoader = new DataLoader<String>() {
+    @Override
+    public String load(String id) throws IOException {
+      return cache.getString(id);
+    }
+  };
+
+  private DataLoader<InputStream> streamServerLoader = new DataLoader<InputStream>() {
+    @Override
+    public InputStream load(String id) throws IOException {
+      InputStream is = client.load(id, REQUEST_METHOD, true, CONNECT_TIMEOUT, getReadTimeout());
+      try {
+        cache.put(id, is);
+      } catch (IOException e) {
+        throw new IllegalStateException("Error saving to WS cache", e);
+      }
+      is.close();
+      return cache.getStream(id);
+    }
+  };
+
+  private DataLoader<InputStream> streamCacheLoader = new DataLoader<InputStream>() {
+    @Override
+    public InputStream load(String id) throws IOException {
+      return cache.getStream(id);
+    }
+  };
+
+  private class NotAvailableException extends Exception {
+    private static final long serialVersionUID = 1L;
+
+    public NotAvailableException(String message) {
+      super(message);
+    }
+
+    public NotAvailableException(Throwable cause) {
+      super(cause);
+    }
+  }
+
+  public WSLoader(LoadStrategy strategy, PersistentCache cache, ServerClient client, UserProperties settings) {
     this.defautLoadStrategy = strategy;
+    this.userProperties = settings;
     this.serverStatus = UNKNOWN;
     this.cache = cache;
     this.client = client;
   }
 
+  private int getReadTimeout() {
+    return userProperties.properties().containsKey(SONAR_WS_TIMEOUT_PROPS) ? Integer.parseInt(userProperties.property(SONAR_WS_TIMEOUT_PROPS)) * 1000 : DEFAULT_READ_TIMEOUT;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS864)